File: Data Structures & Algorithms/subsets/submission-3.java
class Solution {

    public void dfs(int i, ArrayList<Integer> lst, int[] nums, List<List<Integer>> acc)
    {
        if(i == nums.length)
        {
            acc.add(new ArrayList<Integer>(lst));
            return;
        }
        lst.add(nums[i]);
        dfs(i+1, lst, nums, acc);
        lst.remove(lst.size() - 1);
        dfs(i+1, lst, nums, acc);
        
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ret = new ArrayList<>();
        ArrayList<Integer> lst = new ArrayList<Integer>();
        dfs(0, lst, nums, ret);
        return ret;

    }
}

File: Data Structures & Algorithms/subsets-ii/submission-3.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        subset = []
        ret = []
        def dfs(i):
            if i >= len(nums):
                ret.append(subset.copy())
                return
            subset.append(nums[i])
            dfs(i+1)
            subset.pop()
            i += 1
            while i < len(nums) and nums[i] == nums[i-1]:
                i += 1
            if i <= len(nums):
                dfs(i)
        dfs(0)
        return ret